apps/src/blockly/addons/cdoUtils.js

```diff
@@ -106,9 +109,168 @@ export function getUserTheme(themeOption) {
 }
 
 export function getCode(workspace) {
+  // Begin collecting data about our ability to correctly serialize a workspace in JSON format.
+  if (
+    experiments.isEnabled(experiments.BLOCKLY_JSON) ||
+    Math.random() <= 0.05
+  ) {
+    testJsonSerialization(workspace);
+  }
   return Blockly.Xml.domToText(Blockly.Xml.blockSpaceToDom(workspace));
-  // After supporting JSON block sources, change to:
-  // return JSON.stringify(Blockly.serialization.workspaces.save(workspace));
+}
+
+function testJsonSerialization(workspace) {
+  const FIREHOSE_STUDY = 'blockly-json';
+  const FIREHOSE_EVENT_PASS = 'blocks-identical';
+  const FIREHOSE_EVENT_FAIL = 'differences-found';
+  Blockly.Events.disable();
+
+  // Create an array of blocks based on JSON serialization of the current workspace.
+  const blockJson = Blockly.serialization.workspaces.save(workspace);
+  const tempJsonWorkspace = new Blockly.Workspace();
+  Blockly.serialization.workspaces.load(blockJson, tempJsonWorkspace);
+  const jsonBlocks = tempJsonWorkspace.getAllBlocks();
+
+  // Create an array of blocks based on the XML encoding of the current workspace.
+  const blockXml = Blockly.Xml.blockSpaceToDom(workspace);
+  const tempXmlWorkspace = new Blockly.Workspace();
+  Blockly.Xml.domToWorkspace(blockXml, tempXmlWorkspace);
+  const xmlBlocks = tempXmlWorkspace.getAllBlocks();
+
+  // compareBlockArrays returns an array of differences found.
+  const differences = compareBlockArrays(xmlBlocks, jsonBlocks);
+  const result =
+    differences.length > 0 ? FIREHOSE_EVENT_FAIL : FIREHOSE_EVENT_PASS;
+  // Log a record to Firehose/Redshift.
+  const recordData = {
+    projectUrl: dashboard.project.getShareUrl(),
+    blocklyVersion: Blockly.blockly_.VERSION,
+    differences: differences,
+  };
+  firehoseClient.putRecord(
+    {
+      study: FIREHOSE_STUDY,
+      event: result,
+      data_json: JSON.stringify(recordData),
+    },
+    {
+      alwaysPut: true, // Allows logging from development environments
+      includeUserId: true,
+    }
+  );
+  if (experiments.isEnabled(experiments.BLOCKLY_JSON)) {
```

Comment:
Can you clarify here: this experiment is going to be released for everyone, right (at least 5% of the time)? But your PR description says, "An experiment flag can also be used... to load it to the console to aid in internal debugging." Does that mean the `experiments.BLOCKLY_JSON` refers to an extra, optional experimental flag on top of the experiment that will be sampling in production?

Reply:
The new utility will run for all users, 5% of the time. If we want to do additional testing ourselves, we can use the experiment flag which will force the utility to run 100% of the time until disabled. Additionally, I've added a DCDO flag as a first check. Unless `DCDO.get('blockly-json-experiment', true)`, we generate or compare blocks for anyone. This should give us some added comfort in case anything unexpected happens (performance impacts to users or overwhelming Firehose).